### PR TITLE
Netperf

### PR DESCRIPTION
**Pull request by tedsta**
Multithreaded netperf benchmark. Launches a bunch of netperf servers and clients (1:1). Added a `--netperf_num_streams` flag to specify the number of threads. Netperf clients are now launched from a script that gets copied over to the client VMs. The servers are launched with a bash for-loop in the prepare phase.

@yuyantingzero @ehankland @cwilkes PTAL

### CODE REVIEW

File: perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py [hunk 1]
```diff
@@ -121,74 +266,85 @@ def RunNetperf(vm, benchmark_name, server_ip):
...
+    # Only 1 netperf thread
+    throughput_sample, latency_samples, histogram = parsed_output[0]
+    return [throughput_sample] + latency_samples
+  else:
+    # Multiple netperf threads
+
+    samples = []
+
+    # Unzip parsed output
+    throughput_samples, _, latency_histograms = [list(t)
```
Comment (noahl) [inline perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py]: It looks like you assume that multiple netperf threads imply latency histograms. What happens if you run multiple threads of `TCP_STREAM`? Is that possible?

File: perfkitbenchmarker/scripts/netperf_test_scripts/netperf_test.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  netperf_cmd = FLAGS.netperf_cmd
+  num_streams = FLAGS.num_streams
+  port_start = FLAGS.port_start
+
+  assert(netperf_cmd)
+  assert(num_streams >= 1)
+  assert(port_start)
+
+  stdouts = [None for _ in range(num_streams)]
```
Comment (noahl) [inline perfkitbenchmarker/scripts/netperf_test_scripts/netperf_test.py]: You can also write this as `[None] * num_streams`, although I'm not sure which version I prefer. Up to you.

File: tests/linux_benchmarks/netperf_benchmark_test.py [hunk 3]
```diff
@@ -50,6 +52,21 @@ def _ConfigureIpTypes(self, run_external=True, run_internal=True):
...
+        histogram, [0, 20, 30, 74, 80, 100])
+    self.assertEqual(stats['p0'], 1)
+    self.assertEqual(stats['p20'], 1)
+    self.assertEqual(stats['p30'], 2)
+    self.assertEqual(stats['p74'], 2)
+    self.assertEqual(stats['p80'], 5)
+    self.assertEqual(stats['p100'], 5)
+    self.assertTrue(abs(stats['stddev'] - 1.538) <= 0.001)
+
+  def testParseNetperfOutput(self):
```
Comment (noahl) [inline tests/linux_benchmarks/netperf_benchmark_test.py]: Remove this test? Appears to be a stub.

File: perfkitbenchmarker/network.py [hunk 4]
```diff
@@ -51,12 +51,14 @@ def GetFirewall(cls):
         benchmark_spec.firewalls[key] = cls()
       return benchmark_spec.firewalls[key]
 
-  def AllowPort(self, vm, port):
+  def AllowPort(self, vm, start_port, end_port=None):
```
Comment (noahl) [inline perfkitbenchmarker/network.py]: In this and a couple other places, you seem to have the `start_port` and `end_port` changes that you already submitted in a different PR. Merging the master branch into your branch should fix this.

File: perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py [hunk 5]
```diff
@@ -121,74 +266,85 @@ def RunNetperf(vm, benchmark_name, server_ip):
...
+                                                 for t in zip(*parsed_output)]
+    # They should all have the same units
+    throughput_unit = throughput_samples[0].unit
+    # Extract the throughput values from the samples
+    throughputs = [s.value for s in throughput_samples]
+    # Compute some stats on the throughput values
+    throughput_stats = sample.PercentileCalculator(throughputs, [50, 90, 99])
```
Comment (cwilkes) [inline perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py]: Since we import numpy we might want to look at numpy.percentile http://docs.scipy.org/doc/numpy-1.8.1/reference/generated/numpy.percentile.html to do this calculation .  It doesn't take a histogram but rather an array of the values.

File: perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py [hunk 6]
```diff
@@ -93,25 +106,157 @@ def Prepare(benchmark_spec):
...
-    vms[1].AllowPort(DATA_PORT)
+    # Open all of the command and data ports
+    vms[1].AllowPort(PORT_START, PORT_START + num_streams * 2 - 1)
+  netserver_cmd = ('for i in $(seq {port_start} 2 {port_end}); do '
+                   '{netserver_path} -p $i & done').format(
+                       port_start=PORT_START,
+                       port_end=PORT_START + num_streams * 2 - 1,
+                       netserver_path=netperf.NETSERVER_PATH)
+  vms[1].RemoteCommand(netserver_cmd)
+  # vms[1].RemoteCommand('%s -p %s' % (netperf.NETSERVER_PATH, COMMAND_PORT))
```
Comment (tedsta) [inline perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py]: Whoops!

File: perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py [hunk 7]
```diff
@@ -93,25 +106,157 @@ def Prepare(benchmark_spec):
...
+
+  # Install some stuff on the client vm
+  vms[0].Install('pip')
+  vms[0].RemoteCommand('sudo pip install python-gflags==2.0')
+
+  # Copy remote test script to client
+  scratch_dir = vms[0].GetScratchDir()
+  vms[0].RemoteCommand('sudo mkdir -p %s/run/' % scratch_dir)
+  vms[0].RemoteCommand('sudo chmod 777 %s/run/' % scratch_dir)
+  for file_name in REMOTE_SCRIPT_FILES:
```
Comment (tedsta) [inline perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py]: I kept it for flexibility, yeah. It would be good to have some first-class support for remote scripts since we're using them in multiple benchmarks now. I opted to just remove the unnecessary logic for now, though.

File: perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py [hunk 8]
```diff
@@ -93,25 +106,157 @@ def Prepare(benchmark_spec):
...
+  banner = next(fp)
+  assert banner.startswith('MIGRATED'), stdout
+  r = csv.DictReader(fp)
+  results = next(r)
+  logging.info('Netperf Results: %s', results)
+  assert 'Throughput' in results
+
+  # Create the throughput sample
+  throughput = float(results['Throughput'])
+  unit = {'Trans/s': TRANSACTIONS_PER_SECOND,  # *_RR
```
Comment (tedsta) [inline perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py]: Done

File: perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py [hunk 9]
```diff
@@ -93,25 +106,157 @@ def Prepare(benchmark_spec):
...
+    metric = '%s_Throughput' % benchmark_name
+  else:
+    metric = '%s_Transaction_Rate' % benchmark_name
+  meta_keys = [('Confidence Iterations Run', 'confidence_iter'),
+               ('Throughput Confidence Width (%)', 'confidence_width_percent')]
+  metadata.update({meta_key: results[np_key] for np_key, meta_key in meta_keys})
+  throughput_sample = sample.Sample(metric, throughput, unit, metadata)
+
+  # No tail latency for throughput.
+  if unit == MBPS:
```
Comment (tedsta) [inline perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py]: I also don't like this if-statement. It was how I found it. Now that I think about it, the logic around `enable_latency_histograms` eliminates the need for this if-statement. Removed!

File: perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py (hunk 1 above)
Comment (tedsta) [inline perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py]: Yes, running `TCP_STREAM` with multiple threads is possible. In that case, `enable_latency_histograms` will be false and the `latency_histograms`, which will be None in this case, is ignored.

File: perfkitbenchmarker/network.py (hunk 4 above)
Comment (tedsta) [inline perfkitbenchmarker/network.py]: I actually did rebase off of master after that PR merged, but I realized that I forgot to change the `BaseVirtualMachine` and `BaseFirewall` APIs. I figured it'd be alright to slip that into this PR, but if you want I can make a separate PR.

File: perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py (hunk 5 above)
Comment (tedsta) [inline perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py]: I was thinking about this too. PercentileCalculator also computes other statistics than just percentiles. `stddev`, for example. At the very least it ought to be renamed. That's a matter for another PR though, IMO.

File: tests/linux_benchmarks/netperf_benchmark_test.py (hunk 3 above)
Comment (tedsta) [inline tests/linux_benchmarks/netperf_benchmark_test.py]: Done

**Comment by tedsta** [comment]
@noahl Made some changes to address your comments

File: perfkitbenchmarker/network.py (hunk 4 above)
Comment (noahl) [inline perfkitbenchmarker/network.py]: Keeping it in this PR is fine with me.

**Comment by noahl** [comment]
@tedsta Nice work! LGTM
